### DISCUSSION TITLE
CCA bug?

### DISCUSSION THREAD

**Discussion by edutamorim** [vegandevs/vegan#506] [Q&A]
Hi, recently I've been working on a canonical correspondence analysis (CCA) and I'm fixing a bug.
In my matrix of environmental variables, I have 9 columns in total.

when I apply the cca(x,y) function, it only analyzes the first 3 variables.

I tried using some adaptations like cca(x, [,4:9]), cca(x, [,5:9]), thus limiting which columns it would use to perform the analysis. And yet only the first 3 variables are read.... Can anyone help?
![image](https://user-images.githubusercontent.com/101893148/159068965-a3c49bc7-e42a-4c37-bdaa-035705c10d92.png)
![image](https://user-images.githubusercontent.com/101893148/159069018-a4181b9b-19dd-454b-a578-472553559ab4.png)

**Comment by gavinsimpson** [comment]
Your VIFs look so large as to indicate that some of your 9 variables are essentially perfectly linearly correlated with one or a linear combination of other variables.

This suggests that the rank of your environmental data matrix is much lower than 9.

In other words, you don't have 9 independent variables, instead having 3 that are not linear combinations of the others.

**Comment by edutamorim** [comment]
Hi Gavin, thanks so much for reply.  Indeed, these variables present a high correlation.

However, it sounds very strange to me, because when I analyze the last 4 variables [,6:9], or intermediate variables, [, 3:8], the CCA continues using only the first 3 variables. 

And yet, when I use other statistical programs, all variables are plotted on the graph.

However, I understand your comment. And again, thank you very much for the help.

**Comment by jarioksa** [comment]
What was the exact command you used? The one in your example (`cca(x, [,4:9])`) is illegal and won't work.

**Comment by jarioksa** [comment]
We have absolutely no information to answer your questions. The VIFs are high-ish, but I expect you to get `NA` values with aliased variables (and even with lower VIFs: aliased & removed variables do not influence VIFs, but they are only based on remainging variables). You  aslo get an informative message in the output to warn about this:
```
Some constraints or conditions were aliased because they were redundant
```
Aliasing means that some variables are copies of other variables or linear combinations of other variables. This will also be reflected in printed output which gives you to the ranks of constrained and residual unconstrained components. In case of aliasing, the constrained rank is lower than the number of columns in your model matrix. That rank also gives you the number of axes calculated, and it is the number of independent dimensions you can have in your data.

**Comment by miguelsali** [comment]
Hi, edutamorim.

I’ve had the same problem, exactly as you described. Out of my 6 variables, only the first 3 are used for the CCA analysis.
Did you find a solution that you could share?
What other statistical programs did you use to perform your CCA analysis?
Thank you very much in advance.

**Reply by gavinsimpson** [reply]
@miguelsali unless you rewrite the rules of linear algebra, what you are asking for doesn't exist. The same thing would affect a linear model fitted with `lm()`.

The variables you have included in the model are rank deficient; they provide fewer unique pieces of information than the number of columns the add to the model matrix. Put another way, some of your model terms are essentially completely explained by one or more of the other terms in the model and hence cannot be uniquely identified and are thus dropped (aliased).

You could consider dropping some of the variables you threw at the model/data in the hope that the things you actually care about remain in the model and are not aliased. Or you could consider adding the aliased variables into the ordination post-hoc with `envfit()`.
